scripts/flowctl.sh

```diff
@@ -0,0 +1,208 @@
+            parse_option "--docker-image" DOCKER_IMAGE "consume" $ARGS
+            ;;
+        --debug-script*)
+            parse_option "--debug-script" DEBUG_SCRIPT "consume" $ARGS
+            ;;
+        --command*)
+            parse_option "--command" DOCKER_COMMAND "consume" $ARGS
+            ;;
+    esac
+done
+
+if [[ "${DEBUG_SCRIPT}" != "true" && "${DEBUG_SCRIPT}" != "false" ]]; then
+    log_fatal "The --debug-script option requires a value of true or false. Got '${DEBUG_SCRIPT}'"
+fi
+if [[ "${DEBUG_SCRIPT}" = true ]]; then for (( i=1; i <= "$#"; i++ )); do echo "BARG(${i}): ${!i}"; done; fi
+if [[ "${DEBUG_SCRIPT}" = true ]]; then for key in "${!ARGS[@]}"; do echo "ARG(${key}): ${ARGS[$key]}"; done; fi
+
+# Check that we can find the docker socket
+if [[ ! -e "$DOCKER_SOCK" ]] ; then
+    log_fatal "Could not find the docker socket. You can specifiy the location with --docker-sock=/full/path"
+fi
+
+# Get the docker socket group owner and add that to the container to allow manipulation by flowctl
+if ! DOCKER_SOCK_GID=$(stat -Lc '%g' ${DOCKER_SOCK} 2>/dev/null); then
+    # Try the BSD variant if that fails
+    if ! DOCKER_SOCK_GID=$(stat -Lf '%g' ${DOCKER_SOCK} 2>/dev/null); then
+        echo "WARNING: Could not determine gid of docker socket ${DOCKER_SOCK}."
+        DOCKER_SOCK_GID=""
+    fi
+fi
+if [[ ! -z "$DOCKER_SOCK_GID" ]]; then
+    DOCKER_EXTRA_OPTS+="--group-add ${DOCKER_SOCK_GID} "
+fi
+
+# Make sure FLOWCTL_DIRECTORY exists before mapping
+if [[ ! -e ${FLOWCTL_DIRECTORY} ]]; then
+    if ! mkdir ${FLOWCTL_DIRECTORY}; then
+        log_fatal "Could not create working directory ${FLOWCTL_DIRECTORY}"
+    fi
+fi
+
+# Provide the full mapping to the source file if specified and make sure the workdir is set to the path
+if [[ ! -z "${FLOWCTL_SOURCE}" ]]; then
+    FLOWCTL_SOURCE_DIR="$(realpath $(dirname ${FLOWCTL_SOURCE}))"
+    DOCKER_EXTRA_OPTS+="-v ${FLOWCTL_SOURCE_DIR}:${FLOWCTL_SOURCE_DIR} -w ${FLOWCTL_SOURCE_DIR} "
+fi
+
+# If the port option was specified map the same port outside of the container
+if [[ ! -z "${FLOWCTL_PORT}" ]]; then
+    DOCKER_EXTRA_OPTS+="-p ${FLOWCTL_PORT}:${FLOWCTL_PORT} "
+fi
+
+# This is a workaround on MacOS because it requires root in order to run docker-in-docker
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    # Even though the container is running as root, files created on mapped volumes in MacOS still maintin the users ownership
+    DOCKER_UID="0"
+
+    # Attempt to use docker in qemu (assuming it's supported) until we can more accurately work with multiple architectures
+    if [[ `uname -m` == 'arm64' ]]; then
+        DOCKER_EXTRA_OPTS+="--platform linux/amd64 "
+    fi
+fi
+
+# Build docker command
+CMD="${DOCKER_EXEC} run -it --rm \
+--user ${DOCKER_UID} \
+-v ${FLOWCTL_DIRECTORY}:${FLOWCTL_CONTAINAER_DIRECTORY} \
+-v ${DOCKER_SOCK}:/var/run/docker.sock \
+--network ${FLOWCTL_NETWORK} \
+${DOCKER_EXTRA_OPTS} \
+-v /var/tmp:/var/tmp -e TMPDIR=/var/tmp \
```

Comment:
It might be worth adding a comment here about why the tempdir needs to be the same path on the host and inside the container, since it's caused some confusion in the past.

Reply:
Sure thing. I added explanation of the user as well.